File: F#/chap07/Chap0702.fs
namespace chap0702

open Chap0701

type OrderId = Undefined

module C0702a =
    type Order =
        { OrderId: OrderId
          IsValidated: bool
          IsPriced: bool
          AmountToBill: decimal option }

type CustomerInfo = Undefined
type Address = Undefined
type ValidatedOrderLine = Undefined

type ValidatedOrder =
    { OrderId: OrderId
      CustomerInfo: CustomerInfo
      BillingAddress: Address
      ShippingAddress: Address
      OrderLines: ValidatedOrderLine list }

type PricedOrderLine = Undefined
type BillingAmount = Undefined

type PricedOrder =
    { OrderId: OrderId
      CustomerInfo: CustomerInfo
      BillingAddress: Address
      ShippingAddress: Address
      OrderLines: PricedOrderLine list
      AmountToBill: BillingAmount }

type Order =
    | Unvalidated of UnvalidatedOrder
    | Validated of ValidatedOrder
    | Priced of PricedOrder

File: Scala/chap07/src/main/scala/Chap0702.scala
import Chap0702.OrderId

object Chap0702:
  case class OrderId(value: Any)

  case class CustomerInfo(value: Any)

  case class Address(value: Any)

  case class ValidatedOrderLine(value: Any)

  case class ValidatedOrder(
                             orderId: OrderId,
                             customerInfo: CustomerInfo,
                             shippingAddress: Address,
                             billingAddress: Address,
                             orderLines: Seq[ValidatedOrderLine],
                           )

  case class PricedOrderLine(value: Any)

  case class BillingAmount(value: Any)

  case class PricedOrder(
                          orderId: OrderId,
                          customerInfo: CustomerInfo,
                          shippingAddress: Address,
                          billingAddress: Address,
                          orderLines: Seq[PricedOrderLine],
                          amountToBill: BillingAmount,
                        )

  case class UnvalidatedOrder(value: Any)

  enum Order:
    case Unvalidated(value: UnvalidatedOrder)
    case Validated(value: ValidatedOrder)
    case Priced(value: PricedOrder)

object C0702a:
  case class Order(
                    orderId: OrderId,
                    isValidated: Boolean,
                    isPriced: Boolean,
                    AmountToBill: Option[BigDecimal],
                  )
